kartothek/io_components/metapartition.py

```diff
@@ -1249,14 +1249,31 @@ def partition_on(self, partition_on):
                 raise ValueError(
                     "Trying to `partition_on` on a column with an explicit index!"
                 )
-        new_mp = self.as_sentinel().copy(partition_keys=partition_on)
 
         if isinstance(partition_on, str):
             partition_on = [partition_on]
         partition_on = self._ensure_compatible_partitioning(partition_on)
 
         new_data = self._partition_data(partition_on)
 
+        # find a nice origin for the schema
+        if new_data:
```

Comment:
Why do we need this complicated logic? The partitioning shouldn't change the schema and the label / uuid of the original schema should be enough to track the partitions down.

If this is necessary, please add a comment what it is supposed to be doing. The split/zip/join/if is a bit confusing
Consider refactoring it into it's own function.

Reply:
Regarding your top level comment
> please look at the new test case which illustrates the behavior (imo) quite well.

The new test case yields a violation

```
Schema violation
	
	Origin schema: core/P=2/uuid2
	Origin reference: core/P=1/uuid1
	
	Diff:
```

but there is a one-to-one mapping between P values and UUIDs in this example so it doesn't help me to understand.

I don't understand why we have to go through this label transformation logic here. To be exact, I don't even understand what this is doing. `new_data`s keys should be `P=1/uuid1` for the first partition and `P=2/uuid1` for the second partition. In this test case, the new labels are exactly what you would want to print. Is this logic about the missing tables in the keys?
In a more elaborate test case where the partition_on split actually splits the data (e.g. with P values 1 and 2) we would receive
`P=1/uuid1`
and
`P=2/uuid1`
which are again what you probably want to have (excluding tables)

1. Have you thought about attaching the origin to a given schema during MetaPartition initialization? The origin seems to be always connected to the label of the `MetaPartition` and by implementing the attachment of the origin there we don't have to go through this transformation. Plus this might safe us from inconsistencies in any other transformations where one might change labels in the future.
2. My original question is rather: why does the full label add any more value than e.g. `uuid1`? `uuid1` should determine the entirety of the input data

Reply:
UUIDs are random IDs by default (UUID4), the tests just hardcodes them for better error message assertion. So they are not that useful to determine which parts of your dataset have conflicting schemas. Sure, you can open a storage explorer, dig through the files with the UUIDs (which you only find by listing the whole dataset store, because no central metadata file will exist due to the canceled write transaction or the central metadata file does not contain the new files due to a canceled update transaction). I would very much call this a bad UX and therefore a bug.

So what I am trying here is to give the user more hints where the data is. You can only have 1 origin per MetaPartition instance, but normally, this is good enough, because for most usages 1 MetaPartition only spawns a tiny part of the dataset, often even exactly on partition-on combination. In case (see test) it spawns more (Q in P-Q in the test), we can still try to give a more helpful origin.

Reply:
I got the UX part. Just wanted to make sure I didn't miss anything.

Back to point 1)
Wouldn't is be possible to do something like:

```python
class MetaPartition:
    def __init__(..., label=None, table_meta=None, ...):
        self.label = label or gen_uuid()
        self.table_meta = {
            table: schema.with_origin(self.label) 
            for table, schema in table_meta.items()
        }
```

Reply:
This would work if the partition-on only creates a single new partition, otherwise `MetaPartition.table_meta` is 1 instance for all MPs within the nested/container MP. If you look at the test-case, this is exactly what happens for Q. You get 2 Q-values for 1 nested MP (containing 2 sub-MPs each). Another solution could be to store not 1 table_meta for all (sub)MPs within the container/nested MP, but to keep all of them. In that case, we could also have an origin for all these schemas 🤔

Reply:
Ok, took my a while but I finally understand what the test case and the code is supposed to be doing... (I hope, correct me if I'm wrong)
The exception shows `core/P=2/uuid2` but the partition label/path would be `core/P=2/Q=1/uuid2` *and* `core/P=2/Q=2/uuid2`, wouldn't it? I.e. the Q part is filtered out?

If this is the case, we should probably add all physical origins to the schemawrapper. As you said youself, iterating over the blob is *not* recommended and `core/P=2/<>/uuid2` is not sufficient to look up the file.
So we'd rather require something like

```
schema_wrapper.add_origins(sorted(new_data.keys()))
```

Yes, No?

Reply:
> The exception shows core/P=2/uuid2 but the partition label/path would be core/P=2/Q=1/uuid2 and core/P=2/Q=2/uuid2, wouldn't it? I.e. the Q part is filtered out?

Exactly. Sorry for not writing more docs for the specific test case.

> If this is the case, we should probably add all physical origins to the schemawrapper. 

Nice idea, didn't think of that. Will do so. :)

Reply:
If this is the case my proposed solution with `__init__` would of course not work out of the box. One could add a loop over the `table_meta` dict in `add_metapartition`, though (No preference, whatever is easier and consistent)